### ISSUE TITLE
Mysqli Iterator fails unit tests

### ISSUE THREAD

**Issue by photodude** [joomla-framework/database#55]
The Mysqli Iterator fails the unit tests. Off hand from the failures, it seems there may be an indexing issue adding an additional row to the arrays.

There were 4 failures:
```php
1) Joomla\Database\Tests\IteratorMysqliTest::testForEach with data set #0 ('title', '#__dbtest', null, 'stdClass', 0, 0, array(stdClass Object (...), stdClass Object (...), stdClass Object (...), stdClass Object (...)), null)
136
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     0 => stdClass Object (...)
     1 => stdClass Object (...)
     2 => stdClass Object (...)
     3 => stdClass Object (...)
+    4 => stdClass Object (...)
 )
/database/Tests/IteratorMysqliTest.php:137

2) Joomla\Database\Tests\IteratorMysqliTest::testForEach with data set #2 ('title', '#__dbtest', null, 'stdClass', 20, 2, array(stdClass Object (...), stdClass Object (...)), null)
136
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     0 => stdClass Object (...)
     1 => stdClass Object (...)
+    2 => stdClass Object (...)
 )
/database/Tests/IteratorMysqliTest.php:137

3) Joomla\Database\Tests\IteratorMysqliTest::testForEach with data set #3 ('title, id', '#__dbtest', 'title', 'stdClass', 0, 0, array(stdClass Object (...), stdClass Object (...), stdClass Object (...), stdClass Object (...)), null)
136
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'Testing' => stdClass Object (...)
     'Testing2' => stdClass Object (...)
     'Testing3' => stdClass Object (...)
     'Testing4' => stdClass Object (...)
+    'testRollback' => stdClass Object (...)
 )
/database/Tests/IteratorMysqliTest.php:137

4) Joomla\Database\Tests\IteratorMysqliTest::testCount
153
Failed asserting that 5 matches expected 4.
/database/Tests/IteratorMysqliTest.php:154
```

**Comment by mbabker** [comment]
Something is affecting global state and I don't know what it is.  Every failing test class runs fine in isolation but put the whole suite together and it's just busted.

**Comment by photodude** [comment]
I find it interesting that the IteratorMysqliTests and IteratorMysqlTests seem to be exactly the same, but only the mysqlli tests fail. It seems the mysql tests run before the mysqli tests, could that be polluting the global state since they are running against the same DB?

**Comment by mbabker** [comment]
It's definitely a possibility.

**Comment by photodude** [comment]
Guess that's not it. I ran just the mysqli tests with the flag `--filter 'Mysqli'` and these 4 tests still fail. So it's isolated to something in the mysqli tests.

**Comment by mbabker** [comment]
It's something with the driver tests running before the iterator tests.  If you run only the iterator class it's fine.  If you run driver then iterator it fails.  That's as far as I've gotten with it.

**Comment by photodude** [comment]
I was just working back to that combination too. I got that result too. https://travis-ci.org/photodude/database/jobs/160683774

So the question of the hour is "what is the driver doing different in mysqli from mysql that is polluting the global state?"

**Comment by mbabker** [comment]
All I know is it's something to do with the prepared statement changes because that's when it started.

**Comment by photodude** [comment]
I think the test at fault in the driver is "testRollback" the last failure has that line included

**Comment by mbabker** [comment]
It's the entirety of the transactions.  They should be rolling back entirely but obviously a step is getting missed since that object is still in storage.

**Comment by photodude** [comment]
Drats, It's more than just the transactions. I [excluded that from my test](https://travis-ci.org/photodude/database/jobs/160686157) and driver causes the same failures with failure 3 showing `testTitle` rather than `testRollback`

```php
+    'testTitle' => stdClass Object (...)
```

**Comment by photodude** [comment]
From the tests `transactionStart` and `transactionCommit` are use in the two tests that are contributing to the failures. 


Looking at the drivers  `transactionCommit` doesn't seem too different between the drivers

there are some differences that are in play. 
- MySQL is using PdoDriver and calls parent methods
- MySQLi is using DatabaseDriver

So looking at [`MysqliDriver::transactionStart`](https://github.com/joomla-framework/database/blob/master/src/Mysqli/MysqliDriver.php#L875-L908) vs [`PdoDriver::transactionStart`](https://github.com/joomla-framework/database/blob/master/src/Pdo/PdoDriver.php#L776-L796) + [`MysqlDriver::transactionStart`](https://github.com/joomla-framework/database/blob/master/src/Mysql/MysqlDriver.php#L515-L543) there is a lot more going on with mysqli transactionStart

**Comment by photodude** [comment]
in [`MysqliDriver::transactionStart`](https://github.com/joomla-framework/database/blob/master/src/Mysqli/MysqliDriver.php#L875-L908) do we need to reset `autocommit()` to true at the end of the method, or is that handled somewhere else?

**Comment by mbabker** [comment]
If we need to it shouldn't be there, I honestly don't remember why I added it but it's basically a toggle for auto committing of transactions so it seems somewhat backward if you turn it off, start the transaction, then turn it back on mid-transaction.

**Comment by photodude** [comment]
According to what I've read "You do not need to START TRANSACTION after you have turned autocommit off."
but "calling $conn->commit() will NOT automatically set autocommit() back to 'true'" needing explicitly reset it.

I've now tested `reseting `autocommit`](https://travis-ci.org/photodude/database/jobs/160717668) vs [removing `autocommit`](https://travis-ci.org/photodude/database/jobs/160719934)

Based on this, if we remove `autocommit` the test issues are resolved. PR to come.

**Comment by photodude** [comment]
Closed with #57
